File: apps/demo-be-tsex/src/infrastructure/db/user-repo.ts
import { Context, Effect, Layer } from "effect";
import { SqlClient } from "@effect/sql";
import { SqlError } from "@effect/sql/SqlError";
import type { User, CreateUserData } from "../../domain/user.js";
import { ConflictError } from "../../domain/errors.js";
import type { UserStatus } from "../../domain/types.js";

export interface UserRepositoryApi {
  readonly create: (data: CreateUserData) => Effect.Effect<User, ConflictError | SqlError>;
  readonly findByUsername: (username: string) => Effect.Effect<User | null, SqlError>;
  readonly findByEmail: (email: string) => Effect.Effect<User | null, SqlError>;
  readonly findById: (id: string) => Effect.Effect<User | null, SqlError>;
  readonly updateStatus: (id: string, status: UserStatus) => Effect.Effect<void, SqlError>;
  readonly updateDisplayName: (id: string, displayName: string) => Effect.Effect<void, SqlError>;
  readonly updatePassword: (id: string, passwordHash: string) => Effect.Effect<void, SqlError>;
  readonly incrementFailedAttempts: (id: string) => Effect.Effect<void, SqlError>;
  readonly resetFailedAttempts: (id: string) => Effect.Effect<void, SqlError>;
  readonly listUsers: (
    page: number,
    size: number,
    email?: string,
  ) => Effect.Effect<{ items: User[]; total: number }, SqlError>;
}

export class UserRepository extends Context.Tag("UserRepository")<UserRepository, UserRepositoryApi>() {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function rowToUser(row: any): User {
  return {
    id: row.id as string,
    username: row.username as string,
    email: row.email as string,
    passwordHash: row.password_hash as string,
    displayName: row.display_name as string,
    role: row.role as User["role"],
    status: row.status as User["status"],
    failedLoginAttempts: row.failed_login_attempts as number,
    createdAt: new Date(row.created_at as string),
    updatedAt: new Date(row.updated_at as string),
  };
}

export const UserRepositoryLive = Layer.effect(
  UserRepository,
  Effect.gen(function* () {
    const sql = yield* SqlClient.SqlClient;

    return {
      create: (data: CreateUserData) =>
        Effect.gen(function* () {
          const id = crypto.randomUUID();
          const now = new Date().toISOString();
          yield* sql`
            INSERT INTO users (id, username, email, password_hash, display_name, role, status, failed_login_attempts, created_at, updated_at)
            VALUES (${id}, ${data.username}, ${data.email}, ${data.passwordHash}, ${data.displayName}, 'USER', 'ACTIVE', 0, ${now}, ${now})
          `.pipe(
            Effect.catchAll(() =>
              Effect.fail(
                new ConflictError({
                  message: `User with username '${data.username}' or email '${data.email}' already exists`,
                }),
              ),
            ),
          );
          const rows = yield* sql`SELECT * FROM users WHERE id = ${id}`;
          const row = rows[0];
          if (!row) {
            return yield* Effect.fail(new ConflictError({ message: "Failed to create user" }));
          }
          return rowToUser(row);
        }),

      findByUsername: (username: string) =>
        Effect.gen(function* () {
          const rows = yield* sql`SELECT * FROM users WHERE username = ${username}`;
          const row = rows[0];
          return row ? rowToUser(row) : null;
        }),

      findByEmail: (email: string) =>
        Effect.gen(function* () {
          const rows = yield* sql`SELECT * FROM users WHERE email = ${email}`;
          const row = rows[0];
          return row ? rowToUser(row) : null;
        }),

      findById: (id: string) =>
        Effect.gen(function* () {
          const rows = yield* sql`SELECT * FROM users WHERE id = ${id}`;
          const row = rows[0];
          return row ? rowToUser(row) : null;
        }),

      updateStatus: (id: string, status: UserStatus) =>
        Effect.gen(function* () {
          const now = new Date().toISOString();
          yield* sql`UPDATE users SET status = ${status}, updated_at = ${now} WHERE id = ${id}`;
        }),

      updateDisplayName: (id: string, displayName: string) =>
        Effect.gen(function* () {
          const now = new Date().toISOString();
          yield* sql`UPDATE users SET display_name = ${displayName}, updated_at = ${now} WHERE id = ${id}`;
        }),

      updatePassword: (id: string, passwordHash: string) =>
        Effect.gen(function* () {
          const now = new Date().toISOString();
          yield* sql`UPDATE users SET password_hash = ${passwordHash}, updated_at = ${now} WHERE id = ${id}`;
        }),

      incrementFailedAttempts: (id: string) =>
        Effect.gen(function* () {
          const now = new Date().toISOString();
          yield* sql`UPDATE users SET failed_login_attempts = failed_login_attempts + 1, updated_at = ${now} WHERE id = ${id}`;
        }),

      resetFailedAttempts: (id: string) =>
        Effect.gen(function* () {
          const now = new Date().toISOString();
          yield* sql`UPDATE users SET failed_login_attempts = 0, updated_at = ${now} WHERE id = ${id}`;
        }),

      listUsers: (page: number, size: number, email?: string) =>
        Effect.gen(function* () {
          const offset = (page - 1) * size;
          if (email !== undefined && email !== "") {
            const countRows = yield* sql`SELECT COUNT(*) as count FROM users WHERE email LIKE ${"%" + email + "%"}`;
            const total = (countRows[0]?.count as number) ?? 0;
            const rows =
              yield* sql`SELECT * FROM users WHERE email LIKE ${"%" + email + "%"} ORDER BY created_at DESC LIMIT ${size} OFFSET ${offset}`;
            return { items: rows.map(rowToUser), total };
          }
          const countRows = yield* sql`SELECT COUNT(*) as count FROM users`;
          const total = (countRows[0]?.count as number) ?? 0;
          const rows = yield* sql`SELECT * FROM users ORDER BY created_at DESC LIMIT ${size} OFFSET ${offset}`;
          return { items: rows.map(rowToUser), total };
        }),
    };
  }),
);

File: apps/demo-be-rust-axum/src/db/user_repo.rs
use chrono::Utc;
use sqlx::any::AnyRow;
use sqlx::AnyPool;
use uuid::Uuid;

use crate::domain::errors::AppError;
use crate::domain::user::User;

fn row_to_user(row: &AnyRow) -> User {
    use sqlx::Row;
    let id_str: String = row.get("id");
    let created_str: String = row.get("created_at");
    let updated_str: String = row.get("updated_at");
    let deleted_str: Option<String> = row.try_get("deleted_at").ok().flatten();
    let failed_login_attempts: i32 = row.get("failed_login_attempts");
    User {
        id: Uuid::parse_str(&id_str).unwrap_or_else(|_| Uuid::new_v4()),
        username: row.get("username"),
        email: row.get("email"),
        display_name: row.get("display_name"),
        password_hash: row.get("password_hash"),
        role: row.get("role"),
        status: row.get("status"),
        failed_login_attempts: i64::from(failed_login_attempts),
        created_at: chrono::DateTime::parse_from_rfc3339(&created_str)
            .map(|dt| dt.with_timezone(&Utc))
            .unwrap_or_else(|_| Utc::now()),
        created_by: row
            .try_get("created_by")
            .unwrap_or_else(|_| "system".to_string()),
        updated_at: chrono::DateTime::parse_from_rfc3339(&updated_str)
            .map(|dt| dt.with_timezone(&Utc))
            .unwrap_or_else(|_| Utc::now()),
        updated_by: row
            .try_get("updated_by")
            .unwrap_or_else(|_| "system".to_string()),
        deleted_at: deleted_str.as_deref().and_then(|s| {
            chrono::DateTime::parse_from_rfc3339(s)
                .map(|dt| dt.with_timezone(&Utc))
                .ok()
        }),
        deleted_by: row.try_get("deleted_by").ok().flatten(),
    }
}

pub async fn create_user(
    pool: &AnyPool,
    id: Uuid,
    username: &str,
    email: &str,
    display_name: &str,
    password_hash: &str,
    role: &str,
) -> Result<User, AppError> {
    let now_str = Utc::now().to_rfc3339();
    let id_str = id.to_string();

    sqlx::query(
        r#"INSERT INTO users (id, username, email, display_name, password_hash, role, status, failed_login_attempts, created_at, created_by, updated_at, updated_by)
           VALUES ($1, $2, $3, $4, $5, $6, 'ACTIVE', 0, $7, 'system', $8, 'system')"#,
    )
    .bind(&id_str)
    .bind(username)
    .bind(email)
    .bind(display_name)
    .bind(password_hash)
    .bind(role)
    .bind(&now_str)
    .bind(&now_str)
    .execute(pool)
    .await
    .map_err(|e| {
        if e.to_string().contains("UNIQUE")
            || e.to_string().contains("unique")
            || e.to_string().contains("duplicate key")
        {
            AppError::Conflict {
                message: "Username or email already exists".to_string(),
            }
        } else {
            AppError::Database(e)
        }
    })?;

    find_by_id(pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound {
            entity: "user".to_string(),
        })
}

pub async fn find_by_id(pool: &AnyPool, id: Uuid) -> Result<Option<User>, AppError> {
    let id_str = id.to_string();
    let row = sqlx::query(
        r#"SELECT id, username, email, display_name, password_hash, role, status, failed_login_attempts,
                  created_at, created_by, updated_at, updated_by, deleted_at, deleted_by
           FROM users WHERE id = $1"#,
    )
    .bind(&id_str)
    .fetch_optional(pool)
    .await?;

    Ok(row.as_ref().map(row_to_user))
}

pub async fn find_by_username(pool: &AnyPool, username: &str) -> Result<Option<User>, AppError> {
    let row = sqlx::query(
        r#"SELECT id, username, email, display_name, password_hash, role, status, failed_login_attempts,
                  created_at, created_by, updated_at, updated_by, deleted_at, deleted_by
           FROM users WHERE username = $1"#,
    )
    .bind(username)
    .fetch_optional(pool)
    .await?;

    Ok(row.as_ref().map(row_to_user))
}

pub async fn update_status(pool: &AnyPool, id: Uuid, status: &str) -> Result<(), AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query(
        "UPDATE users SET status = $1, updated_at = $2, updated_by = 'system' WHERE id = $3",
    )
    .bind(status)
    .bind(&now_str)
    .bind(&id_str)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn update_role(pool: &AnyPool, id: Uuid, role: &str) -> Result<(), AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query("UPDATE users SET role = $1, updated_at = $2, updated_by = 'system' WHERE id = $3")
        .bind(role)
        .bind(&now_str)
        .bind(&id_str)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn update_display_name(
    pool: &AnyPool,
    id: Uuid,
    display_name: &str,
) -> Result<User, AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query(
        "UPDATE users SET display_name = $1, updated_at = $2, updated_by = 'system' WHERE id = $3",
    )
    .bind(display_name)
    .bind(&now_str)
    .bind(&id_str)
    .execute(pool)
    .await?;
    find_by_id(pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound {
            entity: "user".to_string(),
        })
}

pub async fn update_password_hash(
    pool: &AnyPool,
    id: Uuid,
    password_hash: &str,
) -> Result<(), AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query(
        "UPDATE users SET password_hash = $1, updated_at = $2, updated_by = 'system' WHERE id = $3",
    )
    .bind(password_hash)
    .bind(&now_str)
    .bind(&id_str)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn increment_failed_attempts(pool: &AnyPool, id: Uuid) -> Result<i64, AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query(
        "UPDATE users SET failed_login_attempts = failed_login_attempts + 1, updated_at = $1, updated_by = 'system' WHERE id = $2",
    )
    .bind(&now_str)
    .bind(&id_str)
    .execute(pool)
    .await?;

    use sqlx::Row;
    let row: AnyRow = sqlx::query("SELECT failed_login_attempts FROM users WHERE id = $1")
        .bind(&id_str)
        .fetch_one(pool)
        .await?;
    let val: i32 = row.get::<i32, _>("failed_login_attempts");
    Ok(i64::from(val))
}

pub async fn reset_failed_attempts(pool: &AnyPool, id: Uuid) -> Result<(), AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query("UPDATE users SET failed_login_attempts = 0, updated_at = $1, updated_by = 'system' WHERE id = $2")
        .bind(&now_str)
        .bind(&id_str)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn set_password_reset_token(
    pool: &AnyPool,
    id: Uuid,
    token: &str,
) -> Result<(), AppError> {
    let id_str = id.to_string();
    let now_str = Utc::now().to_rfc3339();
    sqlx::query("UPDATE users SET password_reset_token = $1, updated_at = $2, updated_by = 'system' WHERE id = $3")
        .bind(token)
        .bind(&now_str)
        .bind(&id_str)
        .execute(pool)
        .await?;
    Ok(())
}

pub struct ListUsersResult {
    pub users: Vec<User>,
    pub total: i64,
}

pub async fn list_users(
    pool: &AnyPool,
    page: i64,
    page_size: i64,
    search_filter: Option<&str>,
) -> Result<ListUsersResult, AppError> {
    let offset = (page - 1) * page_size;

    let (users, total) = if let Some(search) = search_filter {
        let pattern = format!("%{search}%");
        let rows = sqlx::query(
            r#"SELECT id, username, email, display_name, password_hash, role, status, failed_login_attempts,
                      created_at, created_by, updated_at, updated_by, deleted_at, deleted_by
               FROM users WHERE email LIKE $1 OR username LIKE $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3"#,
        )
        .bind(&pattern)
        .bind(page_size)
        .bind(offset)
        .fetch_all(pool)
        .await?;
        let users: Vec<User> = rows.iter().map(row_to_user).collect();

        use sqlx::Row;
        let count_row: AnyRow = sqlx::query(
            "SELECT COUNT(*) as cnt FROM users WHERE email LIKE $1 OR username LIKE $1",
        )
        .bind(&pattern)
        .fetch_one(pool)
        .await?;
        let total: i64 = count_row.get::<i64, _>("cnt");
        (users, total)
    } else {
        let rows = sqlx::query(
            r#"SELECT id, username, email, display_name, password_hash, role, status, failed_login_attempts,
                      created_at, created_by, updated_at, updated_by, deleted_at, deleted_by
               FROM users ORDER BY created_at DESC LIMIT $1 OFFSET $2"#,
        )
        .bind(page_size)
        .bind(offset)
        .fetch_all(pool)
        .await?;
        let users: Vec<User> = rows.iter().map(row_to_user).collect();

        use sqlx::Row;
        let count_row: AnyRow = sqlx::query("SELECT COUNT(*) as cnt FROM users")
            .fetch_one(pool)
            .await?;
        let total: i64 = count_row.get::<i64, _>("cnt");
        (users, total)
    };

    Ok(ListUsersResult { users, total })
}
